### PR TITLE
ARROW-12198: [R] bindings for strptime

### PR DESCRIPTION
**Pull request by AlenkaF**

### CODE REVIEW

File: r/R/compute.R [hunk 1]
```diff
@@ -286,3 +286,8 @@ cast_options <- function(safe = TRUE, ...) {
   )
   modifyList(opts, list(...))
 }
+
+strptime_arrow <- function(..., format, unit){
+  a <- collect_arrays_from_dots(list(...))
```
Comment (nealrichardson) [inline r/R/compute.R:291]: I don't think you want `collect_arrays_from_dots` here. This function exists to support the base R behavior like:

```
> sum(1, 2)
[1] 3
```

But `strptime` doesn't take `...` like that.

File: r/R/compute.R [hunk 2]
```diff
@@ -286,3 +286,8 @@ cast_options <- function(safe = TRUE, ...) {
   )
   modifyList(opts, list(...))
 }
+
+strptime_arrow <- function(..., format, unit){
```
Comment (nealrichardson) [inline r/R/compute.R:290]: I'm not sure how useful this function is since it is a thin wrapper around `call_function()` and we can't set it as an S3 method.. More useful would be to add a version of this in the `nse_funcs` in dplyr-functions.R.

In either case, we should match the `base::strptime()` signature: `function (x, format, tz = "")` with the possible addition of `unit` if that's an Arrow feature. 

Also, should `format` and `unit` have default arguments?

File: r/src/compute.cpp [hunk 3]
```diff
@@ -233,6 +233,12 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
                                      max_replacements);
   }
 
+  if (func_name == "strptime") {
+    using Options = arrow::compute::StrptimeOptions;
```
Comment (nealrichardson) [inline r/src/compute.cpp:237]: Does `StrptimeOptions` have a `Defaults()` method in C++? If so, we should call it.

File: r/tests/testthat/test-Array.R [hunk 4]
```diff
@@ -291,6 +291,17 @@ test_that("Timezone handling in Arrow roundtrip (ARROW-3543)", {
   expect_identical(read_feather(feather_file), df)
 })
 
+test_that("strptime", {
+  # array of strings
+  time_strings <- Array$create(c("2018-10-07 19:04:05", NA))
+  # array of timestamps (doesn't work if tz="" is added!)
```
Comment (nealrichardson) [inline r/tests/testthat/test-Array.R:297]: Why not?

File: r/tests/testthat/test-Array.R [hunk 5]
```diff
@@ -291,6 +291,17 @@ test_that("Timezone handling in Arrow roundtrip (ARROW-3543)", {
...
+  timestamps <- Array$create(c(as.POSIXct("2018-10-07 19:04:05"), NA))
+  # array of parsed timestamps
+  parsed_timestamps <- strptime_arrow(time_strings, format = "%Y-%m-%d %H:%M:%S", unit = TimeUnit$MICRO)
```
Comment (nealrichardson) [inline r/tests/testthat/test-Array.R:300]: `unit` should also take human-friendly strings ("s", "ms", etc.); see how this is done in the `timestamp()` function in type.R.

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Thanks for this! Some questions and suggestions.

File: r/tests/testthat/test-Array.R (hunk 4 above)
Comment (AlenkaF) [reply r/tests/testthat/test-Array.R:297]: I haven't figured out yet how to pass `tz` argument to `strptime_arrow ` wrapper function as it is written now. After going through your comments, making the necessary changes then it will not be an issue I think.

File: r/R/compute.R (hunk 2 above)
Comment (AlenkaF) [reply r/R/compute.R:290]: Got it.
Arrow function uses `format` and `unit` as an FunctionOption if I understand correctly, haven't found `tz` yet.

I think they should have defaults, yes: `format = "%Y-%m-%d %H:%M:%S"` and `unit = TimeUnit$MICRO/2L/"us"`.

File: r/src/compute.cpp (hunk 3 above)
Comment (AlenkaF) [reply r/src/compute.cpp:237]: Do I understand correctly that in `scalar_string.cc` line 1744 would suggest `StrptimeOptions` do not have `Defaults()` in C++?

Comment (nealrichardson) [reply r/src/compute.cpp:237]: Yeah looks like it. I made ARROW-12809 to evaluate whether that's correct, but for the purposes of the PR, it's fine.

File: r/tests/testthat/test-Array.R (hunk 4 above)
Comment (nealrichardson) [reply r/tests/testthat/test-Array.R:297]: You're right that strptime in the C++ library doesn't take a timezone argument. Maybe it is expected that if there is a timezone, it will be encoded in the string and parseable by strptime (with the right format string)? But this gets us into the always tricky area of timezone-aware vs. timezone-naive data. @jorisvandenbossche do you have any thoughts/experience with this code?

Comment (rok) [reply r/tests/testthat/test-Array.R:297]: If I understand correctly [system strptime](https://man7.org/linux/man-pages/man3/strptime.3.html) is used so `%Z` or `%z` would work. E.g.: `2020-01-01 23:23:14 Europe/Amsterdam` would be captured by `format = "%Y-%m-%d %H:%M:%S %Z"`.
Capturing timezones would be great IMO but I would listen to Joris here for sure :).

Comment (jorisvandenbossche) [reply r/tests/testthat/test-Array.R:297]: @rok do you know how the system `strptime` handles such a timezone if it is present? (the docs don't specify that, and the output struct doesn't have an entry for that information)

> Maybe it is expected that if there is a timezone, it will be encoded in the string and parseable by strptime (with the right format string)?

The problem here is that *if* a timezone is recorded in the Timestamp type's `tz` field, then the timestamp value is expected to be in UTC, and not localized to the timezone in question (which is what you get from just parsing the string without the timezone information). So basically that means the timestamp needs to be converted from the specific timezone to UTC (if `strptime` doesn't do that for us). And for now, that's not yet something we have implemented, I think (although at some point we probably should?)

Comment (nealrichardson) [reply r/tests/testthat/test-Array.R:297]: Right. I'm not sure, can you cast `timestamp[ms]` to `timestamp[ms, tz="UTC"]` or whatever (without modifying the values in the array, just to set the tz)?

Comment (rok) [reply r/tests/testthat/test-Array.R:297]: @jorisvandenbossche [it seems that](https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/util/value_parsing.h#L663) we don't really use or pass zone information even if strptime captures it. The following passes:
```
  options.timestamp_parsers = {TimestampParser::MakeStrptime("%Y-%m-%d %H:%M:%S %Z")};
  AssertConversion<TimestampType, int64_t>(type, {"1970-01-01 00:00:00 Etc/GMT+6,1970-01-01 00:00:00 UTC\n"}, {{0}, {0}}, options);
```
So timestamp's timezone is currently ignored and the local time is returned. It might be good to document this or even block `%z` and `%Z` to avoid surprises?

Comment (jorisvandenbossche) [reply r/tests/testthat/test-Array.R:297]: > I'm not sure, can you cast `timestamp[ms]` to `timestamp[ms, tz="UTC"]` or whatever (without modifying the values in the array, just to set the tz)?

Casting actually works, but it's simply setting the tz and not changing the actual values, so it's not necessarily the behaviour you would expect (the behaviour would be correct if you assume the tz-naive data to be in UTC, but that seems a wrong assumption).

> So timestamp's timezone is currently ignored and the local time is returned. It might be good to document this or even block `%z` and `%Z` to avoid surprises?

Indeed, it seems we now simply ignore any timezone information in `strptime`:

```
>>> pc.strptime(["2012-01-01 01:02:03+01:00"], format="%Y-%m-%d %H:%M:%S%z", unit="s")
<pyarrow.lib.TimestampArray object at 0x7fad84855220>
[
  2012-01-01 01:02:03
]
>>> pc.strptime(["2012-01-01 01:02:03+01:00"], format="%Y-%m-%d %H:%M:%S%Z", unit="s").type
TimestampType(timestamp[s])
```

I can see some value in keeping that working, so you can at least parse strings that include such information (otherwise you would always get an error with arrow, or you would need to do some string preprocessing to be able to pass them to `strptime`). But then we certainly need to document that. 
On the other hand, if we want to support it in the future, that would change behaviour and erroring now might then be better ..

It seems that at least some `strptime` implementation support `%z` offsets, and store that in `tm->gmt_offset`, which we currently don't use (https://code.woboq.org/userspace/glibc/time/strptime_l.c.html#776). 
At least supporting fixed offsets (`%z`) seems doable (and the result could then be a timestamp type with `tz="UTC"`), properly supporting `%Z` timezone names will be harder.

Comment (rok) [reply r/tests/testthat/test-Array.R:297]: I've created [ARROW-12820](https://issues.apache.org/jira/browse/ARROW-12820) and referenced this discussion.
In context of this issue we could leave a reference to ARROW-12820 in the tests and postpone the timezone functionally?

File: r/R/compute.R (hunk 2 above)
Comment (AlenkaF) [reply r/R/compute.R:290]: @nealrichardson I have trouble with calling `strptime()` function from `nse_funcs` - possible name collision with base. Am I missing something? Thank you!

As for defaults I correct myself, `format` shouldn't have default argument - to match `base::strptime()` signature.

Comment (nealrichardson) [reply r/R/compute.R:290]: What I was suggesting was

```r
nse_funcs$strptime <- function(x, format = "%Y-%m-%d %H:%M:%S", tz = "", unit = "ms") {

}
```

following the model of the other functions there that build Expressions. And if `tz` is not supported somehow, we `stop()` if `tz` is provided.

**Comment by rok** [comment]
@kszucs - could you you enable @AlenkaF to build on CI please?

**Comment by AlenkaF** [comment]
> > @kszucs - could you you enable @AlenkaF to build on CI please?
> 
> Enabled.

Thanks, I guess I need help (approval for CI build) once more =)

**Comment by AlenkaF** [comment]
@jonkeane @nealrichardson - in [failed test](https://github.com/apache/arrow/pull/10334/checks?check_run_id=2655705142#step:8:14756) `expect_equivalent()` fails due to time zone being stored in `lubridate` timestamp but should ignore attributes if I understand correctly. I am running out of ideas how to solve this one :(

Thank you for the help!

**Comment by jonkeane** [comment]
Oh, this was a fun one dig through and figure out what was going on. As I'm sure you've seen, the failure is only in the devel build, and _it turns out_ [that `all.equal.POSIXt()`](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/all.equal.R#L492-L524) has [changed recently](https://github.com/wch/r-source/blob/trunk/doc/NEWS.Rd#L735-L739): https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17277 is the bug report that adds timezone checking. 

Interestingly, [comment 4](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17277#c4) seems to indicate `check.attributes` should be respected, thought it is not (currently) being respected. I've sent an email to the r-devel list and asking if `check.attributes` is supposed to be ignored in this case (but you can see a replication of this in the code below).

For now, I think if you add `check.tzone = FALSE` to the `expect_equivalent()` calls that will fix this (and not cause problems in other versions).

```r
> all.equal(
  list(lubridate::ymd_hms("2018-10-07 19:04:05", tz = NULL)),
  list(lubridate::ymd_hms("2018-10-07 19:04:05")),
  check.attributes = FALSE
)
[1] "Component 1: 'tzone' attributes are inconsistent ('' and 'UTC')"
> all.equal(
  list(lubridate::ymd_hms("2018-10-07 19:04:05", tz = NULL)),
  list(lubridate::ymd_hms("2018-10-07 19:04:05")),
  check.tzone = FALSE
)
[1] TRUE
> 
```

**Comment by AlenkaF** [comment]
Thank you @jonkeane for your feedback.
I corrected the code and would like to ask you for a review, if I may.

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 6]
```diff
@@ -493,3 +493,81 @@ test_that("edge cases in string detection and replacement", {
...
+      collect(),
+    t_stamp,
+    check.tzone = FALSE
+  )
+
+  expect_equivalent(
+    t_string %>%
+      Table$create() %>%
+      mutate(
+        x = strptime(x, format = "%Y-%m-%d %H:%M:%S", unit = TimeUnit$NANO)
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-string-functions.R:529]: Would it be possible to change this to:

```suggestion
        x = strptime(x, format = "%Y-%m-%d %H:%M:%S", unit = "ns")
```

So it's a little closer to what we expect a person would use in practice?

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 7]
```diff
@@ -493,3 +493,81 @@ test_that("edge cases in string detection and replacement", {
...
+      Table$create() %>%
+      mutate(
+        x = strptime(x, format = "%Y-%m-%d %H:%M:%S")
+      ) %>%
+      collect(),
+    t_stamp,
+    check.tzone = FALSE
+  )
+
+  expect_equivalent(
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-string-functions.R:525]: Do you know if these would be equal? I'm not super familiar with how this precision is measured/handled in lubridate/R.

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 8]
```diff
@@ -493,3 +493,81 @@ test_that("edge cases in string detection and replacement", {
...
+    t_string %>%
+      Table$create() %>%
+      mutate(
+        x = strptime(x, format = "%Y-%m-%d %H:%M:%S", tz="PDT")
+      ) %>%
+      collect(),
+    t_stamp,
+    check.tzone = FALSE
+  )
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-string-functions.R:556]: I'm curious what you're testing with this test. Could you explain a little bit more about the case that it's testing?

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 9]
```diff
@@ -493,3 +493,81 @@ test_that("edge cases in string detection and replacement", {
...
+
+  tstring <- tibble(x = c("08-05-2008", NA))
+  tstamp <- tibble(x = c(lubridate::mdy("08/05/2008"), NA))
+  tstamp[[1]] <- as.POSIXct(tstamp[[1]])
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-string-functions.R:560]: I wonder if it would be clearer to do something like `strptime("08-05-2008", format = "%m-%d-%Y")` to generate the expectation here?

**Review by jonkeane — CHANGES_REQUESTED** [review]
This is looking good! I have a few comments (though mostly about the tests)

Would it be good to test that we error like we think we do when a tz argument is given?

File: r/R/dplyr-functions.R [hunk 10]
```diff
@@ -338,3 +338,24 @@ get_stringr_pattern_options <- function(pattern) {
...
+  # we suggest following format = "%Y-%m-%d %H:%M:%S", unit = MILLI/1L/"ms",
+  # (ARROW-12809)
+
+  # ParseTimestampStrptime currently ignores the timezone information (ARROW-12820).
+  # Stop if tz is provided.
+  if (is.character(tz)) {
+    arrow_not_supported("Time zone argument")
+  }
+
+  t_unit <- make_valid_time_unit(unit,c("s" = TimeUnit$SECOND, "ms" = TimeUnit$MILLI, "us" = TimeUnit$MICRO, "ns" = TimeUnit$NANO))
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply r/R/dplyr-functions.R:354]: No need to keep them both.
Would add as comment though, to have the mapping visible?

File: r/tests/testthat/test-dplyr-string-functions.R [hunk 11]
```diff
@@ -493,3 +493,81 @@ test_that("edge cases in string detection and replacement", {
     tibble(x = c("ABC"))
   )
 })
+
+test_that("strptime", {
+
+  t_string <- tibble(x = c("2018-10-07 19:04:05", NA))
+  t_stamp <- tibble(x = c(lubridate::ymd_hms("2018-10-07 19:04:05"), NA))
+  t_stampPDT <- tibble(x = c(lubridate::ymd_hms("2018-10-07 19:04:05", tz = "PDT"), NA))
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply r/tests/testthat/test-dplyr-string-functions.R:501]: I missed a typo in code, line 554: `t_stampPDT` should be used instead of `t_stamp`.

But it relates to your [comment 1](https://github.com/apache/arrow/pull/10334/files#r640822991) and [comment 2](https://github.com/apache/arrow/pull/10334#pullrequestreview-670405891).

I added example `t_stampPDT` to the test to see if I get a warning as `tz` agrument is given. I do but then data is pulled into `R`. Test now correctly fails as `lubridate` converts time to match PDT time zone. But then it should `stop()` as Neal [suggested](https://github.com/apache/arrow/pull/10334#discussion_r635544090) but I am not sure I know how to do that.

Adding separate test to check if we error correctly could be something in the lines of:

```
test_that("errors in strptime", {
  # Error when tz is passed

  x <- Expression$field_ref("x")
  expect_error(
    nse_funcs$strptime(x, tz = "PDT"),
    'Time zone argument not supported by Arrow'
  )
})
```

and then lines from [comment](https://github.com/apache/arrow/pull/10334/files#r640822991) are redundant.

File: r/tests/testthat/test-dplyr-string-functions.R (hunk 8 above)
Comment (AlenkaF) [reply r/tests/testthat/test-dplyr-string-functions.R:556]: See reply here https://github.com/apache/arrow/pull/10334/files#r641408098

File: r/R/dplyr-functions.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (jonkeane) [reply r/R/dplyr-functions.R:354]: You can add a comment if you would like, though it's not strictly necessary

File: r/tests/testthat/test-dplyr-string-functions.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (jonkeane) [reply r/tests/testthat/test-dplyr-string-functions.R:501]: I think that that's what we want actually. 

In many other cases where something isn't (yet) supported in Arrow we automatically pull the data into R with a warning (in some circumstances like this). You might have found this already, but the pattern you propose for the test in your comment matches what we do elsewhere https://github.com/apache/arrow/blob/master/r/tests/testthat/test-dplyr-string-functions.R#L360-L369 which is good (comments about that test also have a bit more explanation about what's going on when the data warnings+is pulled in)

File: r/R/dplyr-functions.R [hunk 12]
```diff
@@ -338,3 +338,24 @@ get_stringr_pattern_options <- function(pattern) {
...
+
+  Expression$create("strptime", 
+                     x, 
+                     options = list(
+                      format = format,
```
Comment (nealrichardson) [inline r/R/dplyr-functions.R:359]: nit: indentation seems off here (and you may just be able to do `options = list(format =...)` inline

File: r/tests/testthat/test-dplyr-string-functions.R (hunk 7 above)
Comment (AlenkaF) [reply r/tests/testthat/test-dplyr-string-functions.R:525]: If I use `check.tzone = FALSE` they are equal. Should I use `expect_equal()` instead of `expect_equivalent()` in the test?

**Comment by AlenkaF** [comment]
@jonkeane @nealrichardson I need approval for the check and the code is ready for another review round. Thank you!

File: r/R/dplyr-functions.R [hunk 13]
```diff
@@ -338,3 +338,22 @@ get_stringr_pattern_options <- function(pattern) {
...
+  # Stop if tz is provided.
+  if (is.character(tz)) {
+    arrow_not_supported("Time zone argument")
+  }
+
+  unit <- make_valid_time_unit(unit, c(valid_time64_units, valid_time32_units))
+
+  Expression$create("strptime", 
+                     x, 
+                     options = list(format = format, unit = unit))
```
Comment (jonkeane) [inline r/R/dplyr-functions.R:358]:
```suggestion
  Expression$create("strptime", x, options = list(format = format, unit = unit))
```

**Review by jonkeane — APPROVED** [review]
This looks good to go. Thank you so much or your contribution, and bearing with us as we went through reviews.

**Review by nealrichardson — APPROVED** [review]
